osbuild/util/sbom/spdx.py

```diff
@@ -1,12 +1,83 @@
+from typing import Callable, Dict, List, Optional, Union
 from uuid import uuid4
 
 import osbuild
 import osbuild.util.sbom.model as sbom_model
 import osbuild.util.sbom.spdx2 as spdx2
 
 
+class SpdxLicenseExpressionFactory:
+    """
+    Factory for creating SPDX license expressions from license strings.
+
+    This factory uses the license-expression package to parse license strings and convert them to SPDX license, if
+    possible.
+
+    The factory also keeps track of all extracted licensing information objects that were created during the conversion
+    process. The extracted licensing information objects are stored in a dictionary, where the key is the license
+    reference ID and the value is the ExtractedLicensingInfo object.
+    """
+
+    def __init__(self, license_index_location=None):
+        self._extracted_license_infos: Dict[str, spdx2.ExtractedLicensingInfo] = {}
+        self._spdx_licensing = None
+        try:
+            get_spdx_licensing = self._get_spdx_licensing()
+            if license_index_location:
+                self._spdx_licensing = get_spdx_licensing(license_index_location)
+            else:
+                self._spdx_licensing = get_spdx_licensing()
+        except ImportError as e:
+            if license_index_location:
+                raise ValueError("The license-expression package is not available. "
+                                 "Specify the license index location has no effect.") from e
+
+    @staticmethod
+    def _get_spdx_licensing() -> Callable:
+        """
+        Return the get_spdx_licensing function from the license-expression package.
+        This is a helper function to make testing and mocking easier.
+        """
+        # pylint: disable=import-outside-toplevel
+        from license_expression import get_spdx_licensing
+        return get_spdx_licensing
+
+    def _to_extracted_license_info(self, license_str: str) -> spdx2.ExtractedLicensingInfo:
+        eli = spdx2.ExtractedLicensingInfo(license_str)
+        return self._extracted_license_infos.setdefault(eli.license_ref_id, eli)
+
+    def ensure_license_expression(self, license_str: str) -> Union[str, spdx2.ExtractedLicensingInfo]:
+        """
+        Convert a license string to a valid SPDX license expression or wrap it in an ExtractedLicensingInfo object.
+
+        This function uses the license-expression package to parse the license string and convert it to an SPDX license
+        expression. If the license string can't be parsed and converted to an SPDX license expression, it is wrapped in an
+        ExtractedLicensingInfo object.
+
+        If the license-expression package is not available, the license string is always wrapped in an
+        ExtractedLicensingInfo object.
+        """
+        if self._spdx_licensing is None:
+            return self._to_extracted_license_info(license_str)
+
+        # pylint: disable=import-outside-toplevel
+        from license_expression import ExpressionError
```

Comment:
Instead of having two imports outside the top level, would it be cleaner to have the import at the top with an exception handler that sets a variable when failed/succeeded?  Not sure if it's uglier, easier, harder, or the same.

Reply:
I reworked the code to do import only on the top level.